src/Stack/Package.hs

```diff
@@ -1362,6 +1364,57 @@ findOrGenerateCabalFile pkgDir = do
             _:_ -> Left $ PackageMultipleCabalFilesFound pkgDir files
       where hasExtension fp x = FilePath.takeExtension fp == "." ++ x
 
+-- | Searches user path for hpack executables newer than bundled one.
+findUserExecutableHpack :: HasConfig env => Config -> RIO env HpackExecutable
+findUserExecutableHpack config = do
+  menv <- liftIO $ configEnvOverrideSettings config envSettings
+  mHpackPath <- findExecutable menv "hpack"
+  eres <- case mHpackPath of
+    Nothing -> return $ Left "no user installed hpack found."
+    Just hpackPath -> do
+      result <- withEnvOverride menv
+              $ withProc (toFilePath hpackPath) ["--numeric-version"]
+              $ tryAny . readProcessStdout_
+
+      let unexpectedResult got = Left $ T.concat
+              [ "'"
+              , T.pack (toFilePath hpackPath)
+              , " --numeric-version' did not respond with expected value. Got: "
+              , got
+              ]
+
+      return $ case result of
+        Left err -> unexpectedResult $ T.pack (show err)
+        Right bs -> case parseVersionFromString (takeWhile (not . isSpace) (BL8.unpack bs)) of
+          Nothing -> unexpectedResult $ T.pack (BL8.unpack bs)
+          Just ver
+            | ver >= $(mkVersion VERSION_hpack) -> Right hpackPath
+            | otherwise -> Left $ T.concat
+                [ "Installed hpack is older than bundled one. "
+                , T.pack (toFilePath hpackPath)
+                , " is version "
+                , versionText ver
+                , " but bundled one is "
+                , VERSION_hpack
+                , ". using bundled."
+                ]
+
+  case eres of
+    Right hpackPath -> return $ HpackCommand (toFilePath hpackPath)
+    Left err -> do
+      logDebug err
+      return HpackBundled
+
+  where
+    envSettings =
```

Comment:
Should be able to inherit EnvSettings from somewhere or perhaps use default.  At least, `esIncludeLocals` should be `True`, that will actually make a difference.